modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/TransactionsInMultipleThreads.java

```diff
@@ -0,0 +1,481 @@
+
+        cache.removeAll();
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testCrossCacheTransactionInAnotherThread() throws Exception {
+        withAllIsolationsAndConcurrencies(new IgniteClosure<Object, Void>() {
+            @Override public Void apply(Object o) {
+                try {
+                    TransactionsInMultipleThreads.this.crossCacheTransactionInAnotherThread();
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
+                }
+                return null;
+            }
+        });
+    }
+
+    /**
+     * @throws Exception
+     */
+    private void crossCacheTransactionInAnotherThread() throws Exception {
+        Ignite ignite1 = ignite(0);
+        IgniteTransactions transactions = ignite1.transactions();
+
+        IgniteCache<String, Integer> cache = ignite1.getOrCreateCache(cacheConfiguration(null).setName("testCache"));
+        IgniteCache<String, Integer> cache2 = ignite1.getOrCreateCache(cacheConfiguration(null).setName("testCache2"));
+
+        Transaction tx = transactions.txStart(transactionConcurrency, transactionIsolation);
+
+        cache.put("key1", 1);
+        cache2.put("key2", 2);
+
+        tx.stop();
+
+        IgniteInternalFuture<Boolean> fut = GridTestUtils.runAsync(new Callable<Boolean>() {
+            @Override public Boolean call() throws Exception {
+                Assert.assertNull(transactions.tx());
+                Assert.assertEquals(TransactionState.STOPPED, tx.state());
+
+                tx.resume();
+
+                Assert.assertEquals(TransactionState.ACTIVE, tx.state());
+
+                cache.put("key3", 3);
+                cache2.remove("key2");
+
+                tx.commit();
+                return true;
+            }
+        });
+
+        Assert.assertNull(fut.error());
+        try {
+            fut.get();
+        }
+        catch (IgniteCheckedException e) {
+            fail("transaction in another thread failed.Concurrency control=" + transactionConcurrency +
+                ". Isolation level=" + transactionIsolation);
+        }
+
+        Assert.assertEquals(TransactionState.COMMITTED, tx.state());
+        Assert.assertEquals((long)1, (long)cache.get("key1"));
+        Assert.assertEquals((long)3, (long)cache.get("key3"));
+        Assert.assertFalse(cache2.containsKey("key2"));
+
+        cache2.removeAll();
+        cache.removeAll();
+    }
+
+    /**
+     * @throws Exception
+     */
+    public void testCrossCacheTransactionInAnotherThreadContinued() throws Exception {
+        withAllIsolationsAndConcurrencies(new IgniteClosure<Object, Void>() {
+            @Override public Void apply(Object o) {
+                try {
+                    TransactionsInMultipleThreads.this.crossCacheTransactionInAnotherThreadContinued();
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
```

Comment:
May rename "e" -> "ignored"?
If you need to write it to log, please use Logger, otherwise no-operation requered.
New line before "return" required.